### PR TITLE
Collapsible refinement lists

### PR DESCRIPTION
**Pull request by batragaurav**
https://trello.com/c/oHSey6Qz/114-collapsible-refinement-lists

### CODE REVIEW

File: frontend/src/index.html [hunk 1]
```diff
@@ -143,9 +143,53 @@
     <div class="flex flex-1 mx-auto w-full max-w-screen-xl sm:overflow-y-hidden">
       <div id="filter-panel" class="fixed z-20 top-0 open:flex hidden flex-col w-64 max-w-full h-full text-sm bg-white md:static md:z-auto md:flex md:w-1/5 md:max-w-sm">
         <div class="flex-none p-2 text-center text-teal-800 font-bold bg-submarine-500 uppercase md:hidden">Filter</div>
-        <div class="flex-grow p-2 pb-40 pt-0 overflow-x-hidden overflow-y-auto">
-          <div id="language-list"></div>
+        <div class="flex-none p-2 pb-40 pt-0 overflow-x-hidden overflow-y-auto">
```
Comment (anantakrishna) [inline frontend/src/index.html:146]: I'm not sure we need to make it `flex-none`, please consider whether it is the reason behind this.

File: frontend/src/index.html [hunk 2]
```diff
@@ -143,9 +143,53 @@
...
+          <!-- <div id="language-list"></div>
           <div id="sound-quality-list"></div>
+          <div id="duration-menu"></div>
+          <div id="percentage-menu"></div>
+          <div id="category-list"></div>
+          <div id="year-list"></div>
+          <div id="location-list"></div> -->
```
Comment (anantakrishna) [inline frontend/src/index.html:153]: Please never keep commented code in the commit. We can always get the previous stage of the code from GIT, therefore no need keeping it commented. Even git diff would be more descriptive with removed lines.

File: frontend/src/index.html [hunk 3]
```diff
@@ -143,9 +143,53 @@
...
+                  x-ref="container1"
+                  x-bind:style="selected == 1 ? 'max-height: ' + $refs.container1.scrollHeight + 'px' : ''"
+                >
+                  <div>
+                    <div id="language-list"></div>
+                  </div>
+                </div>
+              </li>
+            </ul>
+          </div>
```
Comment (anantakrishna) [inline frontend/src/index.html:173]: Looks like here is too much of nested elements. The structure is not so clear, i.e. the purpose of each element is unclear. As far as I can see, we need only three elements: the root div which would have `x-data`, the header of the filter (which is clickable) and the content of the filter. Just a quick sketch:

```html
<div id="languages-filter" x-data="…">
  <div id="languages-header" @click="…">Languages</div>
  <div id="languages-list"></div>
</div>
```

File: frontend/src/index.html [hunk 4]
```diff
@@ -143,9 +143,53 @@
...
+          <div id="percentage-menu"></div>
+          <div id="category-list"></div>
+          <div id="year-list"></div>
+          <div id="location-list"></div> -->
+
+          <div>
+            <ul>
+              <li x-data="{selected:null}">
+                <button type="button" @click="selected !== 1 ? selected = 1 : selected = null">
+                  <div><span class="">LANGUAGES</span></div>
```
Comment (anantakrishna) [inline frontend/src/index.html:159]: Keep the text in camel case (Languages), we make it uppercase using CSS (see `uppercase` in `algolia.css`).

File: frontend/src/index.html [hunk 5]
```diff
@@ -143,9 +143,53 @@
...
+                    <div id="language-list"></div>
+                  </div>
+                </div>
+              </li>
+            </ul>
+          </div>
+
+          <div>
+            <ul>
+              <li x-data="{selected:null}">
```
Comment (anantakrishna) [inline frontend/src/index.html:177]: To avoid repeating yourself, and also to avoid mixing JS into HTML, extract Alpine component into `collapsible.js` file. See https://github.com/alpinejs/alpine/tree/v2.8.2#x-data, Extract Component Logic. Also see `modal.js` in our project for example.

File: frontend/src/index.html [hunk 6]
```diff
@@ -143,9 +143,53 @@
...
+                <button type="button" @click="selected !== 1 ? selected = 1 : selected = null">
+                  <div><span class="">LANGUAGES</span></div>
+                </button>
+                <div
+                  class="relative max-h-0 overflow-hidden transition-all duration-700"
+                  style=""
+                  x-ref="container1"
+                  x-bind:style="selected == 1 ? 'max-height: ' + $refs.container1.scrollHeight + 'px' : ''"
```
Comment (anantakrishna) [inline frontend/src/index.html:165]: I thought that `x-show` would be enough. Manipulating `max-height` seems unnecessary to me.

File: frontend/src/search.js [hunk 7]
```diff
@@ -85,7 +85,7 @@ search.addWidgets([
   },
   panel({
     templates: {
-      header: 'Location',
+      // header: 'Location',
```
Comment (anantakrishna) [inline frontend/src/search.js:88]: Remove the `panel` wrappers altogether, we don't need them anymore.

File: frontend/src/index.html [hunk 8]
```diff
@@ -143,9 +143,53 @@
...
           <div id="sound-quality-list"></div>
+          <div id="duration-menu"></div>
+          <div id="percentage-menu"></div>
+          <div id="category-list"></div>
+          <div id="year-list"></div>
+          <div id="location-list"></div> -->
+
+          <div>
+            <ul>
+              <li x-data="{selected:null}">
```
Comment (anantakrishna) [inline frontend/src/index.html:157]: I think `expanded` suits better than `selected` for the collapsible component. And the default value tend to be `true`.

File: frontend/src/index.html [hunk 9]
```diff
@@ -143,9 +143,53 @@
...
+                <button type="button" @click="selected !== 1 ? selected = 1 : selected = null">
```
Comment (anantakrishna) [inline frontend/src/index.html:158]: This logic is to go to `x-spread`. And I'd suggest using boolean operation: `expanded = !expanded`.

**Review by anantakrishna — CHANGES_REQUESTED** [review]

File: frontend/src/index.html (hunk 1 above)
Comment (batragaurav) [reply frontend/src/index.html:146]: There was a typo, I will change it back to flex-grow.

File: frontend/src/index.html (hunk 2 above)
Comment (batragaurav) [reply frontend/src/index.html:153]: Sure, I have removed the commented code.

File: frontend/src/index.html (hunk 4 above)
Comment (batragaurav) [reply frontend/src/index.html:159]: This has been changed and uppercase is being used.

File: frontend/src/index.html (hunk 8 above)
Comment (batragaurav) [reply frontend/src/index.html:157]: implemented expanded instead of selected.

File: frontend/src/index.html (hunk 5 above)
Comment (batragaurav) [reply frontend/src/index.html:177]: window.collapsibleComponent = function () {
  return {
    open: false,

    collapseExpand() {
      this.open = !this.open;
    },
  };
};

File: frontend/src/index.html (hunk 6 above)
Comment (batragaurav) [reply frontend/src/index.html:165]: I have done implementation using x-show by calling collapsibleComponent from collapsible.js :

          <div x-data="collapsibleComponent()">
            <div x-on:click="collapseExpand()" class="ais-Panel-header hover:bg-green-100">Languages</div>
            <div id="language-list" x-show="open" x-transition></div>
          </div>

File: frontend/src/search.js (hunk 7 above)
Comment (batragaurav) [reply frontend/src/search.js:88]: Removed the panel wrappers.

File: frontend/src/index.html (hunk 6 above)
Comment (anantakrishna) [reply frontend/src/index.html:165]: Don't include “Component” in the name. That's redundant.

**Comment by anantakrishna** [comment]
Whenever you fix all the issues, just push new commits to this branch, and I'll review them.

File: frontend/src/index.html (hunk 6 above)
Comment (batragaurav) [reply frontend/src/index.html:165]: I kept the name consistent with the name modalComponent().
Are you saying to make it collapsible() 

<div x-data="collapsible()">

**Comment by batragaurav** [comment]
> Whenever you fix all the issues, just push new commits to this branch, and I'll review them.

Sure

File: frontend/src/index.html (hunk 6 above)
Comment (anantakrishna) [reply frontend/src/index.html:165]: Oh right, `modalComponent` is called like this. However the other Alpine components are named without `Component` - `player`, 'searchreaultItem`. Therefore let's use just `collapsible` now. Later we will get rid of this incosistency.

**Comment by batragaurav** [comment]
@anantakrishna , Hari Bol !, I have committed the code as per the feedback.

File: frontend/src/collapsible.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+/*!
+ * sri sri guru gaurangau jayatah
+ */
+
+window.collapsible = function () {
+  return {
+    open: false,
```
Comment (anantakrishna) [inline frontend/src/collapsible.js:7]: `open` is a bit confusing, as it's a verb and a noun. Let's use `expanded`.

File: frontend/src/index.html [hunk 11]
```diff
@@ -143,58 +143,35 @@
...
-                  x-bind:style="selected == 1 ? 'max-height: ' + $refs.container1.scrollHeight + 'px' : ''"
-                >
-                  <div>
-                    <div id="language-list"></div>
-                  </div>
-                </div>
-              </li>
-            </ul>
+        <div class="flex-grow p-2 pb-40 pt-0 overflow-x-hidden overflow-y-auto">
+          <div x-data="collapsible()">
```
Comment (anantakrishna) [inline frontend/src/index.html:147]: Let's add a comment above each of the filters. Otherwise it's not easy to quickly figure out which block it is.

File: frontend/src/index.html [hunk 12]
```diff
@@ -143,58 +143,35 @@
...
+            <div x-on:click="collapseExpand()" class="ais-Panel-header">Languages</div>
```
Comment (anantakrishna) [inline frontend/src/index.html:148]: Since we are not using the InstantSearch's `panel` widget, there is no need using its CSS class `ais-Panel-header`. Instead, style elements directly here via Tailwind CSS classes. Later we will extract the filter block into some template (partial HTML) to avoid repetitions.

File: frontend/src/index.html [hunk 13]
```diff
@@ -143,58 +143,35 @@
...
-                  x-ref="container1"
-                  x-bind:style="selected == 1 ? 'max-height: ' + $refs.container1.scrollHeight + 'px' : ''"
-                >
-                  <div><div id="sound-quality-list"></div></div>
-                </div>
-              </li>
-            </ul>
+          <div x-data="collapsible()">
+            <div x-on:click="collapseExpand()" class="ais-Panel-header">Sound Quality</div>
+            <div id="sound-quality-list" x-show="open"></div>
```
Comment (anantakrishna) [inline frontend/src/index.html:153]: For the nested elements use `x-spread` to concentrate the logic in the JS file. See the audio player component for example. In this case it might be fine to keep the directives in HTML, because there are only two of them. But I'd suggest us to still use `x-spread` for you to learn it and for consistency.

**Review by anantakrishna — CHANGES_REQUESTED** [review]
Haribol! Only few more remarks are there.

File: frontend/src/index.html (hunk 11 above)
Comment (batragaurav) [reply frontend/src/index.html:147]: Sure

File: frontend/src/collapsible.js [hunk 14]
```diff
@@ -0,0 +1,13 @@
+/*!
+ * sri sri guru gaurangau jayatah
+ */
+
+window.collapsible = function () {
+  return {
```
[the remark this answers is not in the document]
Comment (batragaurav) [reply frontend/src/collapsible.js:6]: Done

File: frontend/src/app.js [hunk 15]
```diff
@@ -8,11 +8,11 @@ import tippy from 'tippy.js';
 import 'tippy.js/dist/tippy.css';
 import './algolia.css';
 import './app.css';
+import './collapsible';
 import './modal';
 import './player';
 import { search } from './search';
 import './search-result-item';
-
```
[the remark this answers is not in the document]
Comment (batragaurav) [reply frontend/src/app.js:15]: Done

File: frontend/src/collapsible.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+    open: false,
+
+    collapseExpand() {
```
[the remark this answers is not in the document]
Comment (batragaurav) [reply frontend/src/collapsible.js:9]: Done

File: frontend/src/index.html (hunk 12 above)
Comment (batragaurav) [reply frontend/src/index.html:148]: does that mean, we'll use the same style elements directy that have been used in ais-Panel-header?

File: frontend/src/index.html (hunk 13 above)
Comment (batragaurav) [reply frontend/src/index.html:153]: I am able to use x-spread but while calling the method **toggleCollapse()** , 
the following statement is not being executed":      **this.expanded = !this.expanded;**
 Any ideas?
---------------------------------
window.collapsible = () => {
  return {
    expanded: false,

    toggleCollapse() {
 this.expanded = !this.expanded;
    },
    toggle: {
      '@click': 'toggleCollapse',
    },
  };
};

File: frontend/src/index.html (hunk 12 above)
Comment (anantakrishna) [reply frontend/src/index.html:148]: Yes.

File: frontend/src/index.html (hunk 13 above)
Comment (anantakrishna) [reply frontend/src/index.html:153]: JS seems correct. Can you show me the corresponding HTML?

**Review by batragaurav — COMMENTED** [review]
Hari Bol @anantakrishna, Can you please look into and advise if something is missing in HTML.

All Other points have been completed

File: frontend/src/index.html (hunk 13 above)
Comment (anantakrishna) [reply frontend/src/index.html:153]: Oh initially I missed the point that the method is called, but the inner statement is not executed. HTML is correct.

How do you know that the statement is not executed? Try installing the [Alpine.js devtools extension](https://chrome.google.com/webstore/detail/alpinejs-devtools/fopaemeedckajflibkpifppcankfmbhk) for Chrome and inspect the component's data.

By the way, now there is no `x-show` directive in your code, that is why changing the `expanded` property does not affect the component UI.

Comment (batragaurav) [reply frontend/src/index.html:153]: I installed the same & here is what I can see.
Do we need any other directive?

image

Comment (anantakrishna) [reply frontend/src/index.html:153]: Let's return to my question in the previous comment: How do you know that the statement is not executed?

Comment (batragaurav) [reply frontend/src/index.html:153]: Actually, this.expanded = !this.expanded is executed and value changes but is not making an impact to do collapse and expand the section.

**Comment by anantakrishna** [comment]
I see, that's what I suspected. The problem is that you don't have the
`x-show` detective anymore on the other div. Therefore why would it react?
:) Just use `x-spread` on it and add `x-show` in js.

File: frontend/src/index.html (hunk 13 above)
Comment (batragaurav) [reply frontend/src/index.html:153]: Hari Bol @anantakrishna 

By adding the following, it works well:

class="overflow-hidden" x-bind:class="{'max-h-0': !expanded}"

image

Comment (anantakrishna) [reply frontend/src/index.html:153]: I think `x-show` along with `transition` would be more straightforward in this case. Also, try to use `x-spread` in HTML and define all other directives in the JS file, as you did for the header (toggle).

By the way, it may be a good idea to rename `toggle` to `header` and `toggleCollapse` method to just `toggle`.

Comment (anantakrishna) [reply frontend/src/index.html:153]: After implementing this just push your new commits into the same branch and I will review the new version.

Comment (batragaurav) [reply frontend/src/index.html:153]: done.
a better name would be **searchCriteriaHeader** or a little shorter **searchHeader** to be more specific.

File: frontend/src/collapsible.js [hunk 17]
```diff
@@ -0,0 +1,21 @@
+/*!
+ * sri sri guru gaurangau jayatah
+ */
+
+window.collapsible = () => {
+  return {
```
Comment (anantakrishna) [inline frontend/src/collapsible.js:6]: Here we can shorten the syntax even more because the only clause of the function body is `return`. See [the documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#:~:text=To%20return%20an%20object%20literal%20expression%20requires%20parentheses%20around%20expression) and [`player.js`](https://github.com/PureBhaktiArchive/archive-web-app/blob/b090312f828b6d822d58e4a776ce77ea829116ef/frontend/src/player.js#L7) for example.

File: frontend/src/collapsible.js [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+    expanded: false,
+
+    toggle() {
+      this.expanded = !this.expanded;
+    },
+    itemsList: {
```
Comment (anantakrishna) [inline frontend/src/collapsible.js:12]: We can shorten this to `contents`, because in the context of the `collapsible` component we should not assume what type of content will be there: items or something else.

File: frontend/src/collapsible.js [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+      'x-show'() {
+        return this.expanded;
+      },
```
Comment (anantakrishna) [inline frontend/src/collapsible.js:15]: It is better to use the following syntax:

```suggestion
      'x-show': 'expanded',
```

similar to [`player.js`](https://github.com/PureBhaktiArchive/archive-web-app/blob/b090312f828b6d822d58e4a776ce77ea829116ef/frontend/src/player.js#L224).

The reason is that inside the nested object (`itemsList` in this case), according to the JavaScript syntax `this` refers to that nested object, not the root object. Therefore technically we are not referring the same `expanded` property. But Alpine is doing some black magic behind the scenes and supply proper `this` to the function in the runtime. But the development environment (VSCode) does not know about this magic, therefore it considers that `this` to be a different context. And when you try to use any of the intellisense features of the IDE, like renaming the original property or finding all references or going to the definition, VSCode will consider the property of the nested object as completely separate property. I'm not sure if the explanation is clear, but anyhow let's just accept this imperfection of this world and use string literal in such cases. At least the developer will be alert and will not expect the intellisense to work in this case.

File: frontend/src/collapsible.js [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+      searchHeader: {
```
Comment (anantakrishna) [inline frontend/src/collapsible.js:16]: 1. The `collapsible` component can be used anywhere, not necessarily in the filter panels.
2. The shorter is better. In the context of the component it is enough to say `header`.

Comment (anantakrishna) [reply frontend/src/collapsible.js:16]: Moreover, for some reason this object became a child of `itemsList`. This is improper, it will not work even.

File: frontend/src/index.html [hunk 21]
```diff
@@ -144,13 +144,42 @@
...
         <div class="flex-none p-2 text-center text-teal-800 font-bold bg-submarine-500 uppercase md:hidden">Filter</div>
         <div class="flex-grow p-2 pb-40 pt-0 overflow-x-hidden overflow-y-auto">
-          <div id="language-list"></div>
-          <div id="sound-quality-list"></div>
-          <div id="duration-menu"></div>
-          <div id="percentage-menu"></div>
-          <div id="category-list"></div>
-          <div id="year-list"></div>
-          <div id="location-list"></div>
+          <!-- Search Criteria -->
```
Comment (anantakrishna) [inline frontend/src/index.html:147]: We call it “Filters” in this context. Better to be consistent in the terminology.

```suggestion
          <!-- Filters -->
```

**Review by anantakrishna — CHANGES_REQUESTED** [review]
Haribol! Few technical points are there, please see and fix.

**Comment by batragaurav** [comment]
Hari Bol Prabhu !!
I am committing the changes. Let's take these changes as the final ones because we are losing track of the time while building the website, which is far most important than these minor changes. We will definitely come back to these or other items/issues later during the optimization phase once we have completed the website work.

I am sure you understand my concern!!!

File: frontend/src/collapsible.js (hunk 17 above)
Comment (anantakrishna) [reply frontend/src/collapsible.js:6]: This is left unfixed. It's quite easy to do because VSCode even suggests this:

image

I will add a commit myself then.

**Review by anantakrishna — APPROVED** [review]
Now there are no code quality issues in my opinion. Please see the Trello card for functional issues which I've discovered while testing the changes locally.

**Review by anantakrishna — APPROVED** [review]

**Review by anantakrishna — APPROVED** [review]
I have refactored a bit. Namely, I've introduced the `posthtml-include` plugin to avoid repetitions. Now all the markup of the filter section is extracted into a separate HTML file, and then included in the main HTML as many times as needed. Also, I've inlined the SVG files, that's how we're using them, not via `img` tag. You can see the resulting code if curious.

**Comment by anantakrishna** [comment]
This functionality will be a part of the whole filter panel redesign, therefore closing this PR.
